src/LCCheating/CheatingParticleIDAlgorithm.cc

```diff
@@ -0,0 +1,105 @@
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*this, pPfoList));
+
+    for (const ParticleFlowObject *const pPfo : *pPfoList)
+    {
+        try
+        {
+            const ClusterList clusterList = pPfo->GetClusterList();
+            const TrackList trackList = pPfo->GetTrackList();
+
+            const int nClusters(clusterList.size());
+            const int nTracks(trackList.size());
+
+            if (nTracks > 1) 
+                throw StatusCodeException(STATUS_CODE_FAILURE);
+
+            PandoraContentApi::ParticleFlowObject::Metadata metadata;
+
+            // Case 1) No Clusters or Tracks
+            if (nClusters == 0 && nTracks == 0)
+            {
+                throw StatusCodeException(STATUS_CODE_FAILURE);
+            }
+            // Case 2) One Track Only
+            else if (nClusters == 0 && nTracks == 1)
+            {
+                const Track *const pTrack(trackList.front());
+                const MCParticle *const pMainMCParticle(MCParticleHelper::GetMainMCParticle(pTrack));
+                metadata.m_particleId = pMainMCParticle->GetParticleId();
+            }
+            // Case 3) Clusters With No Tracks
+            else if (nClusters > 0 && nTracks == 0)
+            {
+                const MCParticle *const pMainMCParticle(MCParticleHelper::GetMainMCParticle(&clusterList));
+                metadata.m_particleId = pMainMCParticle->GetParticleId();
+            }
+            // Case 4) Clusters and Tracks
+            else if (nClusters > 0 && nTracks == 1)
+            {
+                const MCParticle *const pMainClustersMCParticle(MCParticleHelper::GetMainMCParticle(&clusterList));
+                const int particleIDCluster(pMainClustersMCParticle->GetParticleId());
+
+                const Track *const pTrack(trackList.front());
+                const MCParticle *const pMainTrackMCParticle(MCParticleHelper::GetMainMCParticle(pTrack));
+                const int particleIDTrack(pMainTrackMCParticle->GetParticleId());
+
+                if (particleIDCluster == particleIDTrack)
+                {
+                    metadata.m_particleId = particleIDCluster;
+                }
+                else if (m_useClusterOverTrackID)
+                {
+                    std::cout << "Conflict between track and cluster best mc particle.  Using cluster mc particle." << std::endl;
+                    metadata.m_particleId = particleIDCluster;
+                }
+                else 
+                {
+                    std::cout << "Conflict between track and cluster best mc particle.  Using track mc particle." << std::endl;
+                    metadata.m_particleId = particleIDTrack;
+                }
+            }
+
+            PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ParticleFlowObject::AlterMetadata(*this, pPfo, metadata));
+        }
+        catch (StatusCodeException &statusCodeException)
```

Comment:
If I'm being very picky, this try/catch block is rather large, so start to lose context for the exception. The main exceptions that can be raised here are in the GetMainMCParticle (and in this case, should correctly skip over Pfo, printing an error) or your nTrack and nCluster quality cuts, above. This can probably stay as it is, but I thought I'd make the comment anyhow.